File: ja/codes/swift/chapter_searching/linear_search.swift
/**
 * File: linear_search.swift
 * Created Time: 2023-01-28
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* 線形探索（配列） */
func linearSearchArray(nums: [Int], target: Int) -> Int {
    // 配列を走査
    for i in nums.indices {
        // 目標要素が見つかったらそのインデックスを返す
        if nums[i] == target {
            return i
        }
    }
    // 目標要素が見つからなければ -1 を返す
    return -1
}

/* 線形探索（連結リスト） */
func linearSearchLinkedList(head: ListNode?, target: Int) -> ListNode? {
    var head = head
    // 連結リストを走査
    while head != nil {
        // 対象ノードが見つかったら、それを返す
        if head?.val == target {
            return head
        }
        head = head?.next
    }
    // 対象ノードが見つからない場合は null を返す
    return nil
}

@main
enum LinearSearch {
    /* Driver Code */
    static func main() {
        let target = 3

        /* 配列で線形探索を行う */
        let nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8]
        let index = linearSearchArray(nums: nums, target: target)
        print("目標要素 3 のインデックス = \(index)")

        /* 連結リストで線形探索を行う */
        let head = ListNode.arrToLinkedList(arr: nums)
        let node = linearSearchLinkedList(head: head, target: target)
        print("目標ノード値 3 に対応するノードオブジェクトは \(node!)")
    }
}

File: ja/codes/ruby/chapter_searching/linear_search.rb
=begin
File: linear_search.rb
Created Time: 2024-04-09
Author: Blue Bean (lonnnnnnner@gmail.com)
=end

require_relative '../utils/list_node'

### 線形探索（配列） ###
def linear_search_array(nums, target)
  # 配列を走査
  for i in 0...nums.length
    return i if nums[i] == target # 目標要素が見つかったらそのインデックスを返す
  end

  -1 # 目標要素が見つからなければ -1 を返す
end

### 線形探索（連結リスト） ###
def linear_search_linkedlist(head, target)
  # 連結リストを走査
  while head
    return head if head.val == target # 対象ノードが見つかったら、それを返す

    head = head.next
  end

  nil # 対象ノードが見つからない場合は None を返す
end

### Driver Code ###
if __FILE__ == $0
  target = 3

  # 配列で線形探索を行う
  nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8]
  index = linear_search_array(nums, target)
  puts "対象要素 3 のインデックス = #{index}"

  # 連結リストで線形探索を行う
  head = arr_to_linked_list(nums)
  node = linear_search_linkedlist(head, target)
  puts "対象ノード値 3 に対応するノードオブジェクトは #{node}"
end
